bftengine/tests/bcstatetransfer/source_selector_test.cpp

```diff
@@ -0,0 +1,241 @@
+
+// Remove current replica
+TEST(source_selector_test, remove_current_replica_sets_the_current_replica_to_no_replica) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_EQ(source_selector.currentReplica(), NO_REPLICA);
+  source_selector.updateSource(kSampleCurrentTimeMs);
+  ASSERT_NE(source_selector.currentReplica(), NO_REPLICA);
+  source_selector.removeCurrentReplica();
+  ASSERT_EQ(source_selector.currentReplica(), NO_REPLICA);
+}
+
+TEST(source_selector_test, removing_the_current_replica_makes_it_non_preferred_as_well) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  source_selector.updateSource(kSampleCurrentTimeMs);
+  const auto currentReplica = source_selector.currentReplica();
+  ASSERT_TRUE(source_selector.isPreferred(currentReplica));
+  source_selector.removeCurrentReplica();
+  ASSERT_FALSE(source_selector.isPreferred(currentReplica));
+}
+
+// Reset tests
+TEST(source_selector_test, on_construction_is_reset_reports_true) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_TRUE(source_selector.isReset());
+}
+
+TEST(source_selector_test, leave_initial_state_when_there_are_any_preferred_replicas) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_TRUE(source_selector.isReset());
+  source_selector.setAllReplicasAsPreferred();
+  ASSERT_FALSE(source_selector.isReset());
+}
+
+TEST(source_selector_test, leave_initial_state_when_the_source_selection_time_is_not_zero) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_TRUE(source_selector.isReset());
+  source_selector.setSourceSelectionTime(kSampleCurrentTimeMs);
+  ASSERT_FALSE(source_selector.isReset());
+}
+
+TEST(source_selector_test, leave_initial_state_when_the_fetching_timestamp_is_not_zero) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_TRUE(source_selector.isReset());
+  source_selector.setFetchingTimeStamp(GL, kSampleCurrentTimeMs);
+  ASSERT_FALSE(source_selector.isReset());
+}
+
+TEST(source_selector_test, reset_removes_preferred_replicas) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_EQ(source_selector.numberOfPreferredReplicas(), 0);
+  source_selector.setAllReplicasAsPreferred();
+  ASSERT_NE(source_selector.numberOfPreferredReplicas(), 0);
+  source_selector.reset();
+  ASSERT_EQ(source_selector.numberOfPreferredReplicas(), 0);
+}
+
+TEST(source_selector_test, reset_removes_current_replica) {
+  auto source_selector = SourceSelector(replicas, kRetransmissionTimeoutMs, kReplicaReplacementTimeoutMs);
+  ASSERT_TRUE(source_selector.currentReplica() == NO_REPLICA && source_selector.isReset());
```

Comment:
I think it would be better to use separate asserts, because if one fails, it will be clearer which one it was.

Reply:
This is a good point, thanks.